test/cuda/test_makeplan.c

```diff
@@ -0,0 +1,194 @@
+#include "finufft_errors.h"
+#ifdef NDEBUG
+#undef NDEBUG
+#include <assert.h>
+#define NDEBUG
+#else
+#include <assert.h>
+#endif
+
+#include <stdint.h>
+#include <stdio.h>
+#include <stdlib.h>
+
+#include <cufinufft.h>
+#include <cuda_runtime.h>
+
+typedef struct {
+    char *p[2];
+} wasteful_pointers;
+
+// hackish way to make allocation failures happen
+wasteful_pointers alloc_remaining_device_mem() {
+    wasteful_pointers a = {NULL, NULL};
+    for (int i = 0; i < 2; ++i) {
+        size_t free, total;
+        cudaMemGetInfo(&free, &total);
+
+        int ier = 1;
+        int iter = 0;
+        while (ier && (iter < 60)) {
+            ier = cudaMalloc((void **)&a.p[i], free - (1 << iter));
+            iter++;
+        }
+    }
+
+    return a;
+}
+
+void free_wasteful_pointers(wasteful_pointers a) {
+    cudaFree(a.p[0]);
+    cudaFree(a.p[1]);
+}
+
+int main() {
+    cufinufftf_plan plan;
+    // defaults. tests should shadow them to override
+    const int dim = 1;
+    const int type = 1;
+    const int iflag = 1;
+    const float tol = 1e-5;
+    const int ntransf = 1;
+    const int64_t N[3] = {10, 20, 15};
+
+    // Dimension failure
+    {
+        const int dim = 0;
+        assert(cufinufftf_makeplan(type, dim, N, iflag, ntransf, tol, &plan, NULL) == FINUFFT_ERR_DIM_NOTVALID);
+    }
+
+    // 1D failure modes
+    {
+        const int dim = 1;
+        const int type = 1;
+        const int ntransf = 1;
+
+        // nice input should succeed
+        assert(cufinufftf_makeplan(type, dim, N, iflag, ntransf, tol, &plan, NULL) == 0);
+        cufinufftf_destroy(plan);
+
+        // Ignore higher dims, even if invalid
+        {
+            int64_t N[3] = {10, 0, 15};
+            assert(cufinufftf_makeplan(type, dim, N, iflag, ntransf, tol, &plan, NULL) == 0);
+            cufinufftf_destroy(plan);
+        }
+
+        {
+            int64_t N[3] = {0, 20, 15};
+            assert(cufinufftf_makeplan(type, dim, N, iflag, ntransf, tol, &plan, NULL) == FINUFFT_ERR_NDATA_NOTVALID);
+        }
+
+        // cufinufft can't handle arrays bigger than INT32_MAX (cufft limitation)
+        {
+            int64_t N[3] = {(int64_t)INT32_MAX + 1, 1, 1};
+            assert(cufinufftf_makeplan(type, dim, N, iflag, ntransf, tol, &plan, NULL) == FINUFFT_ERR_NDATA_NOTVALID);
+        }
+
+        {
+            const int ntransf = 0;
+            assert(cufinufftf_makeplan(type, dim, N, iflag, ntransf, tol, &plan, NULL) == FINUFFT_ERR_NTRANS_NOTVALID);
+        }
+
+        {
+            const int type = 4;
+            assert(cufinufftf_makeplan(type, dim, N, iflag, ntransf, tol, &plan, NULL) == FINUFFT_ERR_TYPE_NOTVALID);
+        }
+        
+        /* { */
```

Comment:
Why not working?

Reply:
It's not 100% reliable, and could cause any other context using that GPU to fail out. I sometimes had issues where the next `cuda` call would fail out randomly as well.

Reply:
All right. I'll look into separating this into another test in a separate PR.